src/tlc_tools/datasets.py

```diff
@@ -0,0 +1,193 @@
+        return image
+
+    def generate_bb_crop(self, image, bbs):
+        """
+        Crop a bounding box from the image.
+
+        Args:
+            image: PIL.Image, the input image.
+            bbs: list, bounding boxes associated with the image.
+
+        Returns:
+            tuple: (cropped image, label) where label is a tensor.
+        """
+        if not bbs:
+            raise ValueError("No bounding boxes found. Check your sampler.")
+
+        random_bb = random.choice(bbs)
+
+        crop = tlc.BBCropInterface.crop(
+            image,
+            random_bb,
+            self.bb_schema,
+            x_max_offset=self.x_max_offset,
+            y_max_offset=self.y_max_offset,
+            y_scale_range=self.y_scale_range,
+            x_scale_range=self.x_scale_range,
+        )
+
+        label = random_bb["label"]
+        return crop, torch.tensor(label, dtype=torch.long)
+
+    def generate_background(self, image, bbs, max_attempts=100):
+        """
+        Generate a background patch from the image.
+
+        Args:
+            image: PIL.Image, the input image.
+            bbs: list, bounding boxes associated with the image.
+
+        Returns:
+            tuple: (background patch, background label) where label is a tensor.
+        """
+        image_width, image_height = image.size
+        bb_factory = tlc.BoundingBox.from_schema(self.bb_schema)
+        gt_boxes_xywh = [
+            bb_factory([bb["x0"], bb["y0"], bb["x1"], bb["y1"]])
+            .to_top_left_xywh()
+            .denormalize(image_width, image_height)
+            for bb in bbs
+        ]
+
+        for i in range(max_attempts):
+            # Generate a random box
+            x = max(
+                min(int(self.random_gen.normalvariate(mu=image_width // 2, sigma=image_width // 6)), image_width - 1), 0
+            )
+            y = max(
+                min(
+                    int(self.random_gen.normalvariate(mu=image_height // 2, sigma=image_height // 6)),
+                    image_height - 1,
+                ),
+                0,
+            )
+            w = max(
+                min(int(self.random_gen.normalvariate(mu=image_width // 8, sigma=image_width // 16)), image_width - x),
+                1,
+            )
+            h = max(
+                min(
+                    int(self.random_gen.normalvariate(mu=image_height // 8, sigma=image_height // 16)), image_height - y
+                ),
+                1,
+            )
+            proposal_box = [x, y, w, h]
+
+            # Ensure the proposed box does not intersect any ground truth boxes
+            if not any(self._intersects(proposal_box, gt_box) for gt_box in gt_boxes_xywh):
+                break
+
+        if i == max_attempts - 1:
+            # Return a 100x100 black square if no valid background patch is found
+            return Image.new("RGB", (100, 100), color=(0, 0, 0)), torch.tensor(self.background_label, dtype=torch.long)
```

Comment:
I feel OK leaving this up to the user to change if it is a problem (added a warning). 
It certainly can happen, but it is quite degenerate and if it happens frequently, the user should really not be sending include_background=True (because their images are in this case filled with bounding boxes, so finding backgrounds is not a good idea.